test/state/requests.cpp

```diff
@@ -63,33 +71,78 @@ std::optional<Requests> collect_deposit_requests(std::span<const TransactionRece
+            // according to EIP-6110. If any check fails, collection is considered failed.
+            auto read_word_as_size = [&](size_t pos) -> std::optional<size_t> {
+                assert(log.data.size() >= pos + 32);
+                // ABI words are big-endian. Ensure the high bytes fit into size_t.
+                const size_t bytes_for_size_t = sizeof(size_t);
+                const size_t leading = 32 - bytes_for_size_t;
+                for (size_t i = 0; i < leading; ++i)
+                    if (log.data[pos + i] != 0)
+                        return std::nullopt;  // too large to fit into size_t
+                size_t v = 0;
+                for (size_t i = leading; i < 32; ++i)
+                    v = (v << 8) | static_cast<unsigned char>(log.data[pos + i]);
+                return v;
+            };
+
+            constexpr size_t WORD = 32;
+            assert(log.data.size() >= WORD * 5);
+
+            // Read the 5 offsets from the head (first 5 words).
+            std::array<size_t, 5> offsets = {};
+            for (size_t i = 0; i < offsets.size(); ++i)
+            {
+                const auto w = read_word_as_size(i * WORD);
+                if (!w)
+                    return std::nullopt;
+                offsets[i] = *w;
+            }
+
+            // Compute expected offsets and lengths (hard-coded from the deposit ABI layout).
+            constexpr size_t DATA_SECTION = WORD * 5;  // where the dynamic data area starts
+            constexpr size_t PUBKEY_OFFSET = DATA_SECTION;
+            constexpr size_t PUBKEY_SIZE = 48;
+            constexpr size_t WITHDRAWAL_OFFSET = PUBKEY_OFFSET + WORD + pad_to_words(PUBKEY_SIZE);
+            constexpr size_t WITHDRAWAL_SIZE = 32;
+            constexpr size_t AMOUNT_OFFSET =
+                WITHDRAWAL_OFFSET + WORD + pad_to_words(WITHDRAWAL_SIZE);
+            constexpr size_t AMOUNT_SIZE = 8;
+            constexpr size_t SIGNATURE_OFFSET = AMOUNT_OFFSET + WORD + pad_to_words(AMOUNT_SIZE);
+            constexpr size_t SIGNATURE_SIZE = 96;
+            constexpr size_t INDEX_OFFSET = SIGNATURE_OFFSET + WORD + pad_to_words(SIGNATURE_SIZE);
+            constexpr size_t INDEX_SIZE = 8;
+
+            // Offsets in the head point to the length-word of each dynamic field.
+            const std::array<size_t, 5> expected_offsets = {
+                PUBKEY_OFFSET, WITHDRAWAL_OFFSET, AMOUNT_OFFSET, SIGNATURE_OFFSET, INDEX_OFFSET};
+
+            if (offsets != expected_offsets)
+                return std::nullopt;  // layout does not match expected EIP-6110 deposit layout
 
+            // Validate sizes of each field.
+            auto validate_size = [&](size_t offset, size_t expected_size) -> bool {
+                const auto size = read_word_as_size(offset);
+                return size.has_value() && (*size == expected_size);
+            };
+            if (!validate_size(PUBKEY_OFFSET, PUBKEY_SIZE) ||
```

Comment:
I'm confused here. Are we validating the compile-time constants here? If yes, convert this to a `static_assert`.

Reply:
nope, the first argument is for lookup in the `log.data`. I'll improve the name to be more natural